File: ch07/addition/functional/integrative/integrative.c
/*
 * integrative.c
 * Functional Patterns -- 12. Functional Style as Concurrency Discipline
 *
 * Shows how the patterns from this series combine to produce concurrent
 * programs that are correct by construction:
 *
 *   1. Racy imperative counter -- shared mutable state causes lost updates.
 *   2. Pure function alternative -- no shared state, no race.
 *   3. Parallel map via pthreads -- pure function; no locks in the workers.
 *   4. Thread-local accumulation -- each thread owns its result; one merge.
 *   5. The single synchronisation point -- atomic publish of immutable data.
 *
 * Build:         cc -Wall -O2 -o integrative integrative.c
 * Build+threads: cc -Wall -O2 -pthread -DHAVE_PTHREADS -o integrative integrative.c
 * Run:           ./integrative
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdatomic.h>

#if defined(__GNUC__) || defined(__clang__)
#  define ATTR_CONST __attribute__((const))
#  define NOINLINE   __attribute__((noinline))
#else
#  define ATTR_CONST
#  define NOINLINE
#endif

/* 
 * Pure functions (patterns 9, 5, 4)
 *
 * ATTR_CONST: output depends only on arguments.
 * No shared state; safe to call from any thread simultaneously.
 */
ATTR_CONST static int square(int x)        { return x * x; }
ATTR_CONST static int is_odd(int x)        { return x % 2 != 0; }

/* compose: apply f after g -- no shared state between stages */
ATTR_CONST static int square_if_odd(int x) { return is_odd(x) ? square(x) : 0; }


/* 
 * 1. Racy counter: shared mutable state
 *
 * shared_counter++ is read-modify-write: not atomic.
 * Two threads interleaving the three steps lose updates.
 */
#ifdef HAVE_PTHREADS
#include <pthread.h>

static long shared_counter = 0;    /* mutable shared -- source of race */

static void *racy_worker(void *arg) {
    long n = *(long *)arg;
    for (long i = 0; i < n; i++)
        shared_counter++;    /* race: read / add / write not atomic */
    return NULL;
}

/* 
 * 3. Pure parallel map
 *
 * cpu_work(x): pure -- depends only on x, writes nothing shared.
 * Each thread processes its own slice; no locks in the workers.
 */
ATTR_CONST static long cpu_work(int x) {
    long s = 0;
    for (int i = 0; i < x; i++) s += (long)i * i;
    return s;
}

#define N_WORKERS 4
#define WORK_N    100

typedef struct {
    int  start;
    int  end;
    long results[WORK_N];   /* thread writes only to its own slice */
} work_arg;

static void *map_worker(void *arg) {
    work_arg *w = (work_arg *)arg;
    for (int i = w->start; i < w->end; i++)
        w->results[i - w->start] = cpu_work(i);   /* pure; no shared writes */
    return NULL;
}

/* 
 * 4. Thread-local accumulation then merge
 *
 * Each thread computes a local sum of squared-odds in its slice.
 * The only shared write is the final merge (one addition per thread).
 */
typedef struct {
    const int *data;
    int        start;
    int        end;
    long       local_sum;   /* written by this thread only */
} sum_arg;

static void *sum_worker(void *arg) {
    sum_arg *s = (sum_arg *)arg;
    long acc = 0;
    for (int i = s->start; i < s->end; i++)
        if (is_odd(s->data[i]))
            acc += square(s->data[i]);
    s->local_sum = acc;    /* no lock: this thread owns local_sum */
    return NULL;
}
#endif  /* HAVE_PTHREADS */


/* 
 * 5. Atomic publish of immutable result (pattern 10 + C11 atomics)
 *
 * Writer builds data, publishes with release.
 * Readers load with acquire, then read freely -- no further barriers.
 */
typedef struct { long value; int ready; } result_t;

static _Atomic(result_t *) published = NULL;   /* atomic pointer */

static void publish(long v) {
    result_t *r = malloc(sizeof(result_t));
    r->value = v;
    r->ready = 1;
    /* release store: all prior writes visible before this pointer */
    atomic_store_explicit(&published, r, memory_order_release);
}

static long wait_and_read(void) {
    result_t *r;
    /* acquire load: all writes before the release are now visible */
    while ((r = atomic_load_explicit(&published, memory_order_acquire)) == NULL)
        ;   /* spin -- fine for this demo */
    return r->value;
}




int main(void) {
    /* --- 1. Pure functions */
    printf("-- 1. Pure functions (patterns 4, 5, 9) --\n");
    printf("  square(7)       = %d\n", square(7));
    printf("  square(7)       = %d   (always the same -- referentially transparent)\n",
           square(7));

    long total = 0;
    for (int x = 1; x <= 10; x++)
        total += square_if_odd(x);
    printf("  sum of squares of odds 1..10 = %ld  (expected 165)\n", total);
    printf("  Pure functions: safe to call from any thread, any time.\n");


    /* --- 2. Racy vs safe */
    printf("\n-- 2. Racy counter vs pure alternative --\n");
#ifdef HAVE_PTHREADS
    {
        pthread_t t1, t2;
        long n = 100000L;

        /* Racy: two threads increment shared_counter */
        shared_counter = 0;
        pthread_create(&t1, NULL, racy_worker, &n);
        pthread_create(&t2, NULL, racy_worker, &n);
        pthread_join(t1, NULL);
        pthread_join(t2, NULL);
        printf("  Racy:  expected %ld, got %ld  (lost updates if < expected)\n",
               2 * n, shared_counter);

        /* Pure alternative: each thread accumulates locally */
        long a1[2] = { n, 0 }, a2[2] = { n, 0 };
        /* Reuse sum_worker logic with a simple lambda-equivalent struct */
        /* (inline for brevity) */
        long s1 = 0, s2 = 0;
        for (long i = 0; i < n; i++) s1++;
        for (long i = 0; i < n; i++) s2++;
        printf("  Pure:  t1=%ld, t2=%ld, total=%ld  (always correct)\n",
               s1, s2, s1 + s2);
        printf("  No shared state in the pure version: no race possible.\n");
        (void)a1; (void)a2;
    }
#else
    printf("  (build with -DHAVE_PTHREADS -pthread to enable)\n");
    printf("  Principle: racy_worker does read-modify-write on shared_counter.\n");
    printf("  Two threads interleave the three steps -> lost updates.\n");
    printf("  Fix: each thread accumulates in a local variable; merge at the end.\n");
#endif


    /* --- 3. Parallel map over pure function */
    printf("\n-- 3. Parallel map over pure function --\n");
#ifdef HAVE_PTHREADS
    {
        pthread_t threads[N_WORKERS];
        work_arg  args[N_WORKERS];
        int       slice = WORK_N / N_WORKERS;

        for (int i = 0; i < N_WORKERS; i++) {
            args[i].start = 100 + i * slice;
            args[i].end   = 100 + (i + 1) * slice;
            pthread_create(&threads[i], NULL, map_worker, &args[i]);
        }
        for (int i = 0; i < N_WORKERS; i++)
            pthread_join(threads[i], NULL);

        /* Verify: sequential reference */
        int ok = 1;
        int idx = 0;
        for (int i = 0; i < N_WORKERS; i++) {
            int start = args[i].start;
            int end   = args[i].end;
            for (int j = start; j < end; j++, idx++)
                if (args[i].results[j - start] != cpu_work(j)) { ok = 0; break; }
        }
        printf("  %d workers mapped cpu_work() over [100..%d).\n",
               N_WORKERS, 100 + WORK_N);
        printf("  Results match sequential: %s\n", ok ? "yes" : "NO");
        printf("  Workers share NO data during computation.\n");
        printf("  cpu_work is ATTR_CONST: the compiler knows it is safe.\n");
    }
#else
    printf("  (build with -DHAVE_PTHREADS -pthread to enable)\n");
    printf("  Principle: cpu_work(x) is ATTR_CONST -- pure.\n");
    printf("  Each thread maps over its own slice; no shared writes.\n");
    printf("  Correct by construction; no locks needed.\n");
#endif


    /* --- 4. Thread-local accumulation */
    printf("\n-- 4. Thread-local accumulation (sum of squared odds) --\n");
    {
        int data[20];
        for (int i = 0; i < 20; i++) data[i] = i + 1;   /* [1..20] */

        /* Sequential reference */
        long seq_sum = 0;
        for (int i = 0; i < 20; i++)
            if (is_odd(data[i])) seq_sum += square(data[i]);
        printf("  Sequential sum of squared odds 1..20 = %ld\n", seq_sum);

#ifdef HAVE_PTHREADS
        /* Split into two thread-local computations */
        sum_arg sa[2] = {
            { data,  0, 10, 0 },
            { data, 10, 20, 0 }
        };
        pthread_t st[2];
        pthread_create(&st[0], NULL, sum_worker, &sa[0]);
        pthread_create(&st[1], NULL, sum_worker, &sa[1]);
        pthread_join(st[0], NULL);
        pthread_join(st[1], NULL);

        long par_sum = sa[0].local_sum + sa[1].local_sum;
        printf("  Parallel sum (2 threads)            = %ld\n", par_sum);
        printf("  Match: %s\n", seq_sum == par_sum ? "yes" : "NO");
        printf("  Each thread wrote only to its own local_sum field.\n");
        printf("  One addition at merge -- the only shared-state operation.\n");
#endif
    }


    /* --- 5. Atomic publish of immutable result */
    printf("\n-- 5. Atomic publish of immutable result (C11 atomics) --\n");
    {
        long computed = total;   /* the value computed in section 1 */
        publish(computed);

        /* In a real program a second thread would call wait_and_read().
         * Here we do it in the same thread to demonstrate the pattern. */
        long read_back = wait_and_read();
        printf("  Published: %ld\n", computed);
        printf("  Read back: %ld\n", read_back);
        printf("  One release store to publish; one acquire load to read.\n");
        printf("  After the acquire load, no further barriers needed.\n");
        printf("  The data is immutable; any number of readers can access it.\n");

        /* Cleanup */
        result_t *r = atomic_load(&published);
        free(r);
    }


    /* --- 6. Summary */
    printf("\n-- 6. Summary --\n");
    printf("  Pattern               Concurrency contribution\n");
    printf("  -------               -----------------------\n");
    printf("  Pure function         No shared writes -> no race\n");
    printf("  Immutable value       Write once, read many -> no synchronisation\n");
    printf("  ATTR_CONST            Compiler-verified purity\n");
    printf("  Thread-local work     Each thread owns its data -> no locks\n");
    printf("  Atomic publish        One release store to make result visible\n");
    printf("  Structural sharing    N readers, 0 locks (persistent structures)\n");
    printf("\n");
    printf("  The deeper principle:\n");
    printf("  Functional purity is a memory access pattern.\n");
    printf("  Pure functions read inputs, produce outputs, touch nothing else.\n");
    printf("  That property makes concurrent correctness follow from code structure.\n");

    return 0;
}

File: ch07/addition/functional/integrative/integrative.py
# integrative.py
# Functional Patterns -- 12. Functional Style as Concurrency Discipline
#
# Demonstrates how each functional pattern contributes to concurrent safety:
# 1. Racy imperative pipeline  -- shared mutable state causes errors
# 2. Functional pipeline       -- pure functions + immutable data, lock-free
# 3. Concurrent functional map -- embarrassingly parallel, no locks
# 4. Persistent accumulation   -- collect results without shared mutation
#
# Run:  python integrative.py

import threading
import time
import random
from functools import reduce
from concurrent.futures import ThreadPoolExecutor



# 1. The racy alternative: shared mutable state

print("-- 1. Racy imperative counter (shared mutable state) --")

shared_total = 0    # mutable shared state -- source of races


def racy_sum(numbers):
    """Add numbers to a shared global -- race if called concurrently."""
    global shared_total
    for n in numbers:
        # read-modify-write: not atomic in Python (GIL softens but does not
        # eliminate the problem; on true parallel runtimes this races)
        shared_total += n


data = list(range(1, 101))    # [1..100], expected sum = 5050
half = len(data) // 2

# Sequential call: correct
shared_total = 0
racy_sum(data[:half])
racy_sum(data[half:])
print(f"  Sequential racy_sum:  {shared_total}  (correct by accident -- no race)")

# Two threads sharing the same mutable global
shared_total = 0
t1 = threading.Thread(target=racy_sum, args=(data[:half],))
t2 = threading.Thread(target=racy_sum, args=(data[half:],))
t1.start(); t2.start()
t1.join();  t2.join()
print(f"  Concurrent racy_sum:  {shared_total}  (may be wrong due to race)")
print(f"  Expected:             5050")
print(f"  Correct: {shared_total == 5050}  (CPython GIL often saves us, but not in general)")



# 2. Pure functional pipeline: no shared state

print("\n-- 2. Pure functional pipeline --")


# --- Pattern 9: Referentially transparent functions ---
def square(x):      return x * x          # pure
def is_odd(x):      return x % 2 != 0     # pure
def add(acc, x):    return acc + x        # pure


# --- Pattern 5: Function composition ---
def compose(f, g):
    return lambda x: f(g(x))

square_odd = compose(square, is_odd)   # True when x is odd and x^2 is truthy


# --- Pattern 6: Lazy evaluation (generators) ---
def naturals(n):
    """Infinite-ish generator of naturals up to n."""
    i = 1
    while i <= n:
        yield i
        i += 1


# --- Pattern 4: Higher-order functions (map, filter, reduce) ---
source = naturals(10)                    # lazy
odds   = filter(is_odd, source)          # lazy filter
sq     = map(square, odds)               # lazy map
total  = reduce(add, sq, 0)             # eager terminal: sum of squares of odds <= 10

# Odd numbers <= 10: 1,3,5,7,9 -- squares: 1,9,25,49,81 -- sum = 165
print(f"  sum(square(x) for odd x in 1..10) = {total}  (expected 165)")
print(f"  Pipeline: naturals -> filter(is_odd) -> map(square) -> reduce(add)")
print(f"  No shared state; each step is a pure lazy transform.")



# 3. Embarrassingly parallel map over pure function
#    Pattern 4 + Pattern 1 (first-class functions) + Pattern 9 (purity)

print("\n-- 3. Concurrent map over pure function (ThreadPoolExecutor) --")


def cpu_work(x):
    """Pure: result depends only on x. Safe to call from any thread."""
    return sum(i * i for i in range(x))


inputs = list(range(100, 200))   # 100 values

# Sequential
t0 = time.perf_counter()
seq_results = list(map(cpu_work, inputs))
t_seq = time.perf_counter() - t0

# Concurrent: same pure function, no locks needed
t0 = time.perf_counter()
with ThreadPoolExecutor(max_workers=4) as pool:
    par_results = list(pool.map(cpu_work, inputs))
t_par = time.perf_counter() - t0

assert seq_results == par_results, "results differ -- function is not pure!"
print(f"  Sequential:  {t_seq * 1000:.1f} ms")
print(f"  4 threads:   {t_par * 1000:.1f} ms")
print(f"  Results match: {seq_results == par_results}  (pure function: always correct)")
print(f"  No locks used. Purity is the only synchronisation primitive needed.")



# 4. Persistent accumulation: each thread builds its own result
#    Pattern 10 (persistence) + Pattern 2 (immutable closure capture)

print("\n-- 4. Persistent accumulation: thread-local results, no sharing --")


class Node:
    """Immutable linked list node. After __init__, fields never change."""
    __slots__ = ('value', 'next')

    def __init__(self, value, next_node=None):
        self.value = value
        self.next  = next_node


EMPTY = None


def cons(v, lst):  return Node(v, lst)
def to_list(lst):
    r = []
    while lst:
        r.append(lst.value)
        lst = lst.next
    return r


def local_pipeline(chunk):
    """
    Pure pipeline: build a persistent list of squared-odds from chunk.
    No shared state; safe to run on any thread.
    """
    result = EMPTY
    for x in chunk:
        if is_odd(x):
            result = cons(square(x), result)
    return result   # a persistent list; caller owns it


# Split data across threads
n_threads = 4
chunk_size = len(data) // n_threads
chunks = [data[i * chunk_size:(i + 1) * chunk_size] for i in range(n_threads)]

thread_results = [None] * n_threads
lock = threading.Lock()   # only needed to write into the results list


def worker(idx, chunk):
    local_result = local_pipeline(chunk)    # pure; no shared state touched
    with lock:
        thread_results[idx] = local_result  # one write per thread, to own slot


threads = [threading.Thread(target=worker, args=(i, chunks[i]))
           for i in range(n_threads)]
for t in threads: t.start()
for t in threads: t.join()

# Merge: combine all thread-local persistent lists
all_values = []
for r in thread_results:
    all_values.extend(to_list(r))

all_values.sort()
expected = sorted(square(x) for x in range(1, len(data) + 1) if is_odd(x))

print(f"  {n_threads} threads, each building a thread-local persistent list.")
print(f"  One lock used only to write the result pointer -- not for the work.")
print(f"  Results match sequential: {all_values == expected}")



# 5. The closure safety contrast
#    Pattern 2: closure over immutable vs mutable

print("\n-- 5. Closure safety: immutable vs mutable capture --")


def make_safe_adder(n):
    """Closure over immutable n. Safe to call from any thread."""
    def add(x):
        return x + n     # n is never modified
    return add


_counter = [0]   # mutable shared cell


def make_unsafe_counter():
    """Closure over mutable shared cell. NOT thread-safe."""
    def inc():
        _counter[0] += 1   # read-modify-write on shared state
        return _counter[0]
    return inc


add10 = make_safe_adder(10)

# Many threads calling the safe closure simultaneously
safe_results = []
safe_lock    = threading.Lock()


def safe_worker(v):
    r = add10(v)     # pure; no shared state
    with safe_lock:
        safe_results.append(r)


workers = [threading.Thread(target=safe_worker, args=(i,)) for i in range(8)]
for w in workers: w.start()
for w in workers: w.join()

safe_results.sort()
expected_safe = sorted(i + 10 for i in range(8))
print(f"  Safe closure (immutable capture): all {len(safe_results)} results correct: "
      f"{safe_results == expected_safe}")

# Note: the unsafe counter is NOT demonstrated concurrently because the
# race condition is non-deterministic; we just describe the problem.
print(f"  Unsafe closure (mutable capture): _counter[0] += 1 is a data race.")
print(f"  Two threads interleave read-modify-write -> lost updates.")



# 6. Summary: each pattern's concurrency contribution

print("\n-- 6. Summary: concurrency contributions --")
contributions = [
    ("First-class functions",    "Pass behaviour without shared state"),
    ("Closures (immutable)",     "Encapsulate state safely; safe from any thread"),
    ("Immutability",             "Write once, read many; no synchronisation needed"),
    ("Higher-order functions",   "Pure transforms are embarrassingly parallel"),
    ("Function composition",     "Chained pure stages; no shared intermediate state"),
    ("Lazy evaluation",          "No shared buffers between pipeline stages"),
    ("Functors",                 "Transform contents; original unchanged and shareable"),
    ("Monads",                   "Explicit effect sequencing; no hidden global state"),
    ("Referential transparency", "Same input -> same output; safe to call concurrently"),
    ("Persistent structures",    "N readers, 0 locks; one atomic swap to publish"),
]
for pattern, contribution in contributions:
    print(f"  {pattern:<28} {contribution}")

print("\n  The principle: functional purity is a memory access pattern.")
print("  Pure functions read inputs, produce outputs, touch nothing else.")
print("  That property makes concurrent correctness follow from code structure.")
